core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala

```diff
@@ -0,0 +1,402 @@
+    leaderMemberId: String,
+    followerMemberId: String,
+    metadata: Array[Byte],
+    version: Int,
+  ): String = {
+    // The leader rejoins and triggers a rebalance.
+    val rejoinLeaderFuture = Future {
+      sendJoinRequest(
+        groupId = "grp",
+        memberId = leaderMemberId,
+        metadata = metadata,
+        version = version.toShort
+      )
+    }
+
+    TestUtils.waitUntilTrue(() => {
+      val described = describeGroups(groupIds = List("grp"))
+      GenericGroupState.PREPARING_REBALANCE.toString == described.head.groupState
+    }, msg = s"The group is not in PREPARING_REBALANCE state.")
+
+    // A new follower with duplicated group instance id joins.
+    val joinNewFollowerResponseData = sendJoinRequest(
+      groupId = "grp",
+      groupInstanceId = "group-instance-id",
+      metadata = metadata,
+      version = version.toShort
+    )
+
+    TestUtils.waitUntilTrue(() => {
+      val described = describeGroups(groupIds = List("grp"))
+      GenericGroupState.COMPLETING_REBALANCE.toString == described.head.groupState
+    }, msg = s"The group is not in COMPLETING_REBALANCE state.")
+
+    // The old follower rejoin request should be fenced.
+    val rejoinFollowerResponseData = sendJoinRequest(
+      groupId = "grp",
+      memberId = followerMemberId,
+      groupInstanceId = "group-instance-id",
+      metadata = metadata,
+      version = version.toShort
+    )
+
+    val rejoinLeaderFutureResponseData = Await.result(rejoinLeaderFuture, Duration.Inf)
+    val newFollowerMemberId = joinNewFollowerResponseData.memberId
+
+    verifyJoinGroupResponseDataEquals(
+      new JoinGroupResponseData()
+        .setGenerationId(3)
+        .setProtocolType(if (version >= 7) "consumer" else null)
+        .setProtocolName("consumer-range")
+        .setLeader(leaderMemberId)
+        .setMemberId(leaderMemberId)
+        .setMembers(List(
+          new JoinGroupResponseMember()
+            .setMemberId(leaderMemberId)
+            .setMetadata(metadata),
+          new JoinGroupResponseMember()
+            .setMemberId(newFollowerMemberId)
+            .setGroupInstanceId("group-instance-id")
+            .setMetadata(metadata)
+        ).asJava),
+      rejoinLeaderFutureResponseData
+    )
+
+    verifyJoinGroupResponseDataEquals(
+      new JoinGroupResponseData()
+        .setGenerationId(3)
+        .setProtocolType(if (version >= 7) "consumer" else null)
+        .setProtocolName("consumer-range")
+        .setLeader(leaderMemberId)
+        .setMemberId(newFollowerMemberId),
+      joinNewFollowerResponseData
+    )
+
+    verifyJoinGroupResponseDataEquals(
+      new JoinGroupResponseData()
+        .setProtocolName(if (version >= 7) null else "")
+        .setMemberId(followerMemberId)
+        .setErrorCode(Errors.FENCED_INSTANCE_ID.code),
+      rejoinFollowerResponseData
+    )
+
+    newFollowerMemberId
+  }
+
+  private def verifyJoinGroupResponseDataEquals(
```

Comment:
The elements in the list of JoinGroupResponseMember doesn't have a determined order. It would be better if I can directly change the `equals` method in JoinGroupResponseData, but I'm worried that it can cause bugs in other places

Reply:
In this case, it would be better to duplicate the structure with `duplicate()` and to sort in place the field(s) with unstable orders. This is a little more robust because we don't have to modify it to support new fields in the future.

Reply:
See this [commit](https://github.com/apache/kafka/commit/9e884bcff5a9cac8b49c4d3b783c7eedf60feff1) for an example.